core/src/main/java/org/apache/calcite/sql/validate/DelegatingScope.java

```diff
@@ -542,6 +543,208 @@ private int worstKind(Path path) {
     }
   }
 
+
+  /**
+   * Converts a table identifier with an ID column into a fully-qualified identifier.
+   * For example, the dept in "select empno from emp natural join dept" may become
+   * "myschema.dept".
+   *
+   * @param identifier SqlTableIdentifierWithID to qualify
+   * @return A qualified identifier, never null
+   */
+  @Override public SqlTableIdentifierWithIDQualified fullyQualify(
+      SqlTableIdentifierWithID identifier) {
+    // This implementation is based on the fullyQualify for regular identifiers above. We skip
+    // anything that only pertains to columns because we are only handling tables.
+    final SqlTableIdentifierWithID previous = identifier;
+    final SqlNameMatcher nameMatcher = validator.catalogReader.nameMatcher();
+    SqlValidatorNamespace fromNs = null;
+    Path fromPath = null;
+    RelDataType fromRowType = null;
+    final ResolvedImpl resolved = new ResolvedImpl();
+    int size = identifier.getNames().size();
+    int i = size - 1;
+    for (; i > 0; i--) {
+      final SqlTableIdentifierWithID prefix = identifier.getComponent(0, i);
+      resolved.clear();
+      resolve(prefix.getNames(), nameMatcher, false, resolved);
+      if (resolved.count() == 1) {
+        final Resolve resolve = resolved.only();
+        fromNs = resolve.namespace;
+        fromPath = resolve.path;
+        fromRowType = resolve.rowType();
+        break;
+      }
+      // Look for a table alias that is the wrong case.
+      if (nameMatcher.isCaseSensitive()) {
+        final SqlNameMatcher liberalMatcher = SqlNameMatchers.liberal();
+        resolved.clear();
+        resolve(prefix.getNames(), liberalMatcher, false, resolved);
+        if (resolved.count() == 1) {
+          final Step lastStep = Util.last(resolved.only().path.steps());
+          throw validator.newValidationError(prefix,
+              RESOURCE.tableNameNotFoundDidYouMean(prefix.toString(),
+                  lastStep.name));
+        }
+      }
+    }
+    if (fromNs == null || fromNs instanceof SchemaNamespace) {
+      final SqlTableIdentifierWithID prefix1 = identifier.skipLast(1);
+      throw validator.newValidationError(prefix1,
+          RESOURCE.tableNameNotFound(prefix1.toString()));
```

Comment:
Fairly big check in here, for a column that is not qualified by a table alias (337). Can we skip this?

Reply:
I assume it's a similar situation to last time, findQualifyingTableNames doesn't work because we're in the outermost scope, so we can just omit the check and immediately throw an error.